### PR TITLE
mini-alloc: a very simple bump allocator

### PR DESCRIPTION
**Pull request by mikebenfield**

### CODE REVIEW

File: mini-alloc/tests/wasm.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+#[global_allocator]
+static ALLOC: mini_alloc::MiniAlloc = mini_alloc::MiniAlloc::INIT;
+
+#[wasm_bindgen_test]
+fn it_works() {
+    let p1 = Vec::<u8>::with_capacity(700);
+    assert!(p1.as_ptr() as usize > 0);
+    let p2 = Vec::<u8>::with_capacity(65536);
+    assert_eq!(p1.as_ptr() as usize + 700, p2.as_ptr() as usize);
```
Comment (rachel-bousfield) [inline mini-alloc/tests/wasm.rs:16]: `p1.as_ptr().add(700) == p2.as_ptr()`

File: mini-alloc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+unsafe impl Sync for MiniAlloc {}
+
+impl MiniAlloc {
+    pub const INIT: Self = MiniAlloc {
+        start: Cell::new(0),
+    };
+}
+
+fn round_up_to_alignment(val: usize, align: usize) -> usize {
+    (val + align - 1) & (-(align as isize) as usize)
```
Comment (rachel-bousfield) [inline mini-alloc/src/lib.rs:38]: `(val + align - 1) & !(align - 1)`

Also let's make it `const` :)

Comment (mikebenfield) [reply mini-alloc/src/lib.rs:38]: `align` is always a power of two in `Layout`:

https://doc.rust-lang.org/std/alloc/struct.Layout.html

But I can add a comment on this function making it clear that `align` must be a power of two.

File: mini-alloc/tests/wasm.rs (hunk 1 above)
Comment (mikebenfield) [reply mini-alloc/tests/wasm.rs:16]: This will require an `unsafe` block. Is that still your preference?

**Comment by mikebenfield** [comment]
Now supports unix and windows, with a completely separate implementation from wasm32. Addressed your other comments as well. 

I'd like to squash this before merging.

**Comment by mikebenfield** [comment]
Oh, also, the only windows testing I've done is `cargo check --target x86_64-pc-windows-gnu`.

File: mini-alloc/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#[derive(Default)]
+pub struct MiniAlloc {
+    start: Cell<usize>,
+}
+
+unsafe impl Sync for MiniAlloc {}
+
+impl MiniAlloc {
+    pub const INIT: Self = MiniAlloc {
+        start: Cell::new(0),
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply mini-alloc/src/lib.rs]: This is what I changed to, but actually is there a particular reason you preferred this? Ideally I'd like to go back to `Cell` to cut down on the amount of `unsafe` littered around.

File: mini-alloc/src/imp.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+/// The WASM page size, or 2^16 bytes.
+const PAGE_SIZE: usize = 1 << 16;
+
+/// Represents the negation of the allocator's bump offset and boundary
+///
+/// We store the negation because we can align the negative offset in fewer
+/// instructions than the positive offset.
+static mut STATE: Option<(NonZero, usize)> = None;
+
+unsafe fn alloc_impl(layout: Layout) -> Option<*mut u8> {
```
Comment (rauljordan) [inline mini-alloc/src/imp.rs:42]: Hi @mikebenfield should we validate `layout.size()` before doing more stuff here? Maybe check it is non-zero or not some excessive value

File: mini-alloc/src/imp.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    let (neg_offset, neg_bound) = STATE.get_or_insert_with(|| {
+        let heap_base = &__heap_base as *const u8 as usize;
+        let bound = PAGE_SIZE * wasm32::memory_size(0);
```
Comment (rauljordan) [inline mini-alloc/src/imp.rs:45]: Each time this occurs, it has to call wasm32::memory_size(0) which I believe won't change throughout the duration of a program. Is it possible to cache this?

**Review by rauljordan — COMMENTED** [review]
Everything looks great! Only had minor comments

File: mini-alloc/src/imp.rs (hunk 5 above)
Comment (mikebenfield) [reply mini-alloc/src/imp.rs:45]: `get_or_insert_with` will only call this closure the first time this is executed. After that `STATE` will contain a `Some` rather than a `None` and the values will just be retrieved.

**Review by rauljordan — APPROVED** [review]

File: mini-alloc/src/imp.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    let neg_aligned = make_aligned(neg_offset.get(), layout.align());
+    let next_neg_offset = neg_aligned.checked_sub(layout.size())?;
+    let bytes_needed = neg_bound.saturating_sub(next_neg_offset);
+    if bytes_needed != 0 {
+        let pages_needed = 1 + (bytes_needed - 1) / PAGE_SIZE;
+        if wasm32::memory_grow(0, pages_needed) == usize::MAX {
+            return None;
+        }
+        *neg_bound -= PAGE_SIZE * pages_needed;
```
Comment (rachel-bousfield) [inline mini-alloc/src/imp.rs]: This overflows if the max number of pages are allocated.
This is why in #89 we start the boundary 1 off.
- #89

**Review by rachel-bousfield — CHANGES_REQUESTED** [review]
I've pushed a commit that adds a test for various edge cases. We're getting close!

Btw don't worry about force pushing :)

**Review by rachel-bousfield — APPROVED** [review]
Looks great! Nice work!
